fearless/ApplicationLayer/Services/WalletBalanceSubscriptionAdapter.swift

```diff
@@ -0,0 +1,470 @@
+        }
+        buildBalance()
+    }
+}
+
+// MARK: - WalletBalanceBuilderProtocol
+
+private protocol WalletBalanceBuilderProtocol {
+    func buildBalance(
+        for accountInfos: [ChainAssetKey: AccountInfo?],
+        _ metaAccounts: [MetaAccountModel],
+        _ chainAssets: [ChainAsset],
+        _ prices: [PriceData]
+    ) -> [MetaAccountId: WalletBalance]?
+}
+
+// MARK: - WalletBalanceBuilder
+
+private final class WalletBalanceBuilder: WalletBalanceBuilderProtocol {
+    func buildBalance(
+        for accountInfos: [ChainAssetKey: AccountInfo?],
+        _ metaAccounts: [MetaAccountModel],
+        _ chainAssets: [ChainAsset],
+        _ prices: [PriceData]
+    ) -> [MetaAccountId: WalletBalance]? {
+        let walletBalanceMap = metaAccounts.reduce(
+            [MetaAccountId: WalletBalance]()
+        ) { (dict, account) -> [MetaAccountId: WalletBalance]? in
+
+            let splitedChainAssets = split(chainAssets, for: account)
+            let enabledChainAssets = splitedChainAssets.enabled
+            let disabledChainAssets = splitedChainAssets.disabled
+
+            let enabledAssetFiatBalanceInfo = countBalance(
+                for: enabledChainAssets,
+                account,
+                accountInfos,
+                prices
+            )
+
+            let disabledAssetFiatBalanceInfo = countBalance(
+                for: disabledChainAssets,
+                account,
+                accountInfos,
+                prices
+            )
+
+            let enabledAssetFiatBalance = enabledAssetFiatBalanceInfo.totalBalance
+            let disabledAssetFiatBalance = disabledAssetFiatBalanceInfo.totalBalance
+            let totalFiatValue = enabledAssetFiatBalance + disabledAssetFiatBalance
+            let dayChangePercent = countDayChangePercent(for: prices)
+            let dayChangeValue = totalFiatValue * dayChangePercent / 100
+            let isLoaded = enabledAssetFiatBalanceInfo.isLoaded && disabledAssetFiatBalanceInfo.isLoaded
+
+            guard isLoaded else {
+                return nil
+            }
+
+            let walletBalance = WalletBalance(
+                totalFiatValue: totalFiatValue,
+                enabledAssetFiatBalance: enabledAssetFiatBalance,
+                dayChangePercent: dayChangePercent,
+                dayChangeValue: dayChangeValue,
+                currency: account.selectedCurrency
+            )
+
+            var dict = dict
+            dict?[account.metaId] = walletBalance
+            return dict
+        }
+
+        return walletBalanceMap
+    }
+
+    private func countDayChangePercent(for prices: [PriceData]) -> Decimal {
+        let fiatDayChanges = prices.compactMap { priceData in
+            priceData.fiatDayChange
+        }
+
+        let dayChange = fiatDayChanges.reduce(Decimal.zero, +)
+        let dayChangePercent = dayChange / Decimal(fiatDayChanges.count)
```

Comment:
incorrect. example where you have 0% change but it's incorrect value:
a: 1x5000usd
b: 100x1usd

a: -10% (-500usd)
b: +10% (+10usd)